core/unify.ml

```diff
@@ -586,6 +609,28 @@ let rec unify' : unify_env -> (datatype * datatype) -> unit =
                          "' with abstract type '"^string_of_datatype t2^"'")))
     | `Application (_, ls), `Application (_, rs) ->
        List.iter2 (fun lt rt -> unify_type_args' rec_env (lt, rt)) ls rs
+    | `Primitive t, `RecursiveApplication a
+    | `RecursiveApplication a, `Primitive t ->
+       raise (Failure
+                (`Msg ("Cannot unify primitive type '"^string_of_datatype (`Primitive t) ^
+                         "' with recursive type '"^ a.r_name ^"'")))
+    | `RecursiveApplication a1, `RecursiveApplication a2 ->
+        let (n1, args1) = (a1.r_unique_name, a1.r_args) in
+        let (n2, args2) = (a2.r_unique_name, a2.r_args) in
+        if n1 <> n2 then
```

Comment:
This means that two named types are only unifiable if they are the same. I would have expected this to fall back to a structural unification if the names don't match. 

The following was valid before, but fails to type-check now:
```
mutual {
  typename T1(a) = [| Nil | Cons: (a,T1(a)) |];
}
mutual {
  typename T2(a) = [| Nil | Cons: (a,T2(a)) |];
}

sig nil : T1(Int)
var nil = Nil;

sig test :(T2(Int)) -> Int
fun test(x) {
  123
}

test(nil)
```
Here, T1 and T2 are the same type with a different name, but we cannot use them interchangeably.

Of course, this is a design decision, so the others should probably chime in on this.

Reply:
Relatedly to previous comments: When did we start treating recursive types/typename in this way?  Is it a change to behavior as of this PR only, or has this been different for a while (e.g. since introducing mutually recursive types) and this PR is just doing the same thing in a cleaner way?  It's not necessarily a problem, but seems like a nontrivial change - in particular it contradicts what the (old old old) documentation says about `typename`, so if we adopt this PR I request we update the documentation at least enough to reflect this.

Reply:
(Incidentally, Frank's example above works in master if we leave out the `mutual` blocks; both types are translated to alpha-equivalent mu types.)

Reply:
Before this PR, the following was already valid:
```
typename T1(a) = [| Nil | Cons: (a,T1(a)) |];
```

However, this just introduced an alias for a \mu type before. Those aliases made writing complicated (recursive) types easier and were carried around to obtain nice error messages, but were completely irrelevant for type-checking; everything was typed structurally.
(My understanding of) the idea of this PR was to be a bit more clever about type-checking: If we carry around the type names for debugging purposes anyways, we can use them as a shortcut for type-checking: If we compare two types and both turn out to be named types with the same name and same type arguments, they must be the equal, and we don't have to do a structural type comparison. One prerequisite for doing so is to give named types proper identifiers (as done in this PR), whereas previously we just used strings with no consideration for scoping, shadowing, etc.


However, my understanding was that we would still keep the structural semantics: If we compare named types and the names don't match or we compare a named type to something different, we would fall back to a structural equality check. This way, the type system would stay structural, but the fact that some types have names would be used in a more disciplined way to increase performance of the type-checker.

My comment here shows that what I thought the new behavior should be (fall back to structural typing) is not what actually happens in this PR. I don't know if this is a bug or if my understanding of how to combine structural and (pseudo-)nominal typing is different from what Simon had in mind.

When adopting my understanding, `newtype` never actually creates new types, just aliases (which we may then use cleverly).

Reply:
> However, my understanding was that we would still keep the structural semantics: If we compare named types and the names don't match or we compare a named type to something different [...]

Yes, this is one of the things we decided to try a year ago (or something like that) when we discussed how speed up the type checking in wake of #359. However, I thought that would be a separate PR on top of this, whose only goal I thought was to make recursion explicit in the syntax.

Reply:
Thanks, @frank-emirch: your understanding is correct, and this is a bug. Thanks very much for catching it - I'll patch it and add a regression test.

Reply:
Just to clarify: in my mind, the biggest issue fixed by this PR is to remove recursive type inlining (i.e., hacking recursive types using mu-types), which led to substantial problems in the programs I was writing in practice. The explicit mutual syntax is handy, and I think a good design decision.

Reply:
Resolved in 4271aba98fb5c54cda9fd5f9daa2240e367b84ad.